### DISCUSSION TITLE
Remix-run + React-Hook-Form + Zod

### DISCUSSION THREAD

**Discussion by enyelsequeira** [react-hook-form/react-hook-form#9004] [Ideas]
Hello all,

I am wondering if there's a way to integrate remix run with react-hook form, I haven't had any luck with it quite yet. any tips or examples?

**Comment by bluebill1049** [comment]
I believe someone made a package on top of react hook form to integrate with remix. I forgot the name tho...

**Reply by bluebill1049** [reply]
i think this is the one: https://github.com/SeasonedSoftware/remix-forms

**Comment by bluebill1049** [comment]
I did a quick try myself and it works for me with the new Form component.

**Reply by enyelsequeira** [reply]
Hey @bluebill1049 nice that’s awesome. Will have to look at it

**Reply by bluebill1049** [reply]
https://twitter.com/HookForm/status/1625621864331743232

**Reply by jrestall** [reply]
We need a way for the new Form component to execute Remix's useSubmit hook instead of submitting the form itself or integrate validation with Remix's Form component. I gave it a go and because RHF submits the form itself, it means redirects and isSubmitting events etc aren't supported since Remix isn't in control of the form submission.

**Reply by bluebill1049** [reply]
react hook form is a client-side focused form validation library, we want to support all frameworks including other libraries, so our solution needs to be generic enough without biased on specific ones.

Here is a proposal to have better support for the remix and potentially other libraries such as SWR/Tan query


- `fetcher` prop will overwrite the default fetch method and let the framework or library handle the request, let me know your feedback we can release another beta to test it out.

```tsx
// remix
const submit = useSubmit()

<Form 
  action="/api"
  method="post"
  fetcher={(data, { event }) => {
    submit(event.currentTarget);
  }}
 />
```

```tsx
// SWR/Tan query
<Form 
  action="/api"
  method="post"
  fetcher={mutation(data)}
 />
```

**Reply by jrestall** [reply]
Thanks @bluebill1049! I tested and a custom fetcher works well! Maybe an issue I noticed though is event.currentTarget is null in the fetcher event prop which I'm not sure is expected or not, so only using target worked for me. `submit(event.target);` but if target is the element that triggered the event then this might not always be okay.

**Reply by bluebill1049** [reply]
> if target is the element that triggered the event then this might not always be okay.

What do you mean by that? It should be form element by default right?

**Reply by jrestall** [reply]
I was just surprised currentTarget was null as I expected that to be the Form and I expected target to be the button that I clicked. https://thisthat.dev/current-target-vs-target/

I see in the code you are just passing the event through so I'm sure it's fine and most likely I misunderstand how the targets work.

Thanks again for making this change, it looks like it'll benefit a lot of libraries.

**Reply by bluebill1049** [reply]
Good to hear, would you like to share an example for remix implementation. I am thinking of sharing that in the doc.

**Reply by cill-i-am** [reply]
@bluebill1049 is the fetcher prop released? I'm on the latest version and don't see it

**Reply by Pouet--** [reply]
@Rocinante89 The fetcher prop has been deleted in the last release [see here](https://github.com/orgs/react-hook-form/discussions/10427)

**Comment by AlemTuzlak** [comment]
I built this lightweight package https://www.npmjs.com/package/remix-hook-form which wraps the react-hook-form package and adds some extra spice over it for remix. Would love for you guys to check it out!

**Reply by AlemTuzlak** [reply]
@bluebill1049 would love to get your feedback on it!
